### PR TITLE
Ruby: JSON flow summaries

### PR DESCRIPTION
**Pull request by hmac**
Add flow summaries for `ActiveSupport::JSON` and various JSON methods. We consider JSON parsing and generation to be taint-preserving.

### CODE REVIEW

File: ruby/ql/test/library-tests/frameworks/json/JsonDataFlow.ql
Comment (erik-krogh) [inline ruby/ql/test/library-tests/frameworks/json/JsonDataFlow.ql:10]:
```suggestion
from DataFlow::PathNode source, DataFlow::PathNode sink, DefaultTaintFlowConf conf
```

Your inline tests are testing for taint-flow, but in this test you are outputting all the data-flows.  
Is that intentional? 
You could add a negative inline expectation about data-flow?

File: ruby/ql/lib/codeql/ruby/frameworks/Json.qll
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/frameworks/Json.qll:18]:
```suggestion
          "json;;Member[JSON].Method[generate,fast_generate,pretty_generate,dump,unparse,fast_unparse];Argument[0];ReturnValue;taint",
```

https://rubyapi.org/3.1/o/json#method-i-pretty_generate

**Review by alexrford — COMMENTED** [review]
There are a few instance methods for `oj` that look relevant for flow from the `self` arg to the return value, e.g. `as_json`, `to_json` etc: https://rubydoc.info/gems/oj/JSON/GenericObject Are these in scope for this PR - I recognize that there's probably a ton of potential methods to model that might not be used much in practice.

**Comment by alexrford** [comment]
> That's a good question. As you say, there's a lot of potential methods we should model. My preference is to tackle the obvious low-hanging fruit here, and then tackle the various JSON gems in full later on, as part of general non-rails library modelling. What do you think?

Sounds like a good approach.

**Review by alexrford — DISMISSED** [review]
The tests need updating, but otherwise LGTM - I'll reapprove when the tests are fixed.

**Review by alexrford — APPROVED** [review]

**Pull request merged**
